codex/sales/reservations.nim

```diff
@@ -233,10 +238,47 @@ proc update*(
   without key =? obj.key, error:
     return failure(error)
 
-  without oldAvailability =? await self.get(key, Availability), err:
-    if err of NotExistsError:
-      let res = await self.updateImpl(obj)
-      # inform subscribers that Availability has been added
+  let lock = self.locks.mgetOrPut(obj.id, newAsyncLock())
+  try:
+    trace "acquiring lock for availibility", availabilityId=obj.id
+    await lock.acquire()
+    trace "got lock for availibility", availabilityId=obj.id
+    without oldAvailability =? await self.get(key, Availability), err:
+      if err of NotExistsError:
+        let res = await self.updateImpl(obj)
+        # inform subscribers that Availability has been added
+        if onAvailabilityAdded =? self.onAvailabilityAdded:
+          # when chronos v4 is implemented, and OnAvailabilityAdded is annotated
+          # with async:(raises:[]), we can remove this try/catch as we know, with
+          # certainty, that nothing will be raised
+          try:
+            await onAvailabilityAdded(obj)
+          except CancelledError as e:
+            raise e
+          except CatchableError as e:
+            # we don't have any insight into types of exceptions that
```

Comment:
I know that this was already part of the original code, but I couldn't resist commenting on this.

I'm usually more in favor of a hard crash (raise Defect) than hiding the error in the logs. The intent here is that the callback should not raise any exceptions, so if it does then we'd like to know about it as soon as possible.

Reply:
Yeah, this is not the scope of the PR ;-) If you want create please issue and I will fix it up in follow up PR...